### PR TITLE
Add multi tenancy capabilities with ecto meta prefix

### PR DESCRIPTION
**Pull request by dreamingechoes**
Connects to #24.

## What does this PR do?

Adds multi tenancy capabilities leveraging the [Ecto.Query prefix](https://hexdocs.pm/ecto/Ecto.Query.html#module-query-prefix).

### Changes made in lib/paper_trail.ex

A new private function `add_prefix` was added into `lib/paper_trail.ex` in order to 
add the new `:prefix` option into the `Version` changeset and be able to store the data
on the tenant specified by the user.

Removes the unnecessary functions `get_version` and `get_versions` in `lib/paper_trail.ex`
and delegate them to the `PaperTrail.VersionQueries` module.

### Changes made in lib/paper_trail/version_queries.ex

The `version_query` private function was redefined to allow the addition of the options
specified (only `:prefix` at this moment) on the filter query.

Some new definitions of the main functions were added to allow the new option parameter,
as well as some specs and basic documentation.

### Other changes

Documentation of the `README.md` file was updated with some information about this
new capability.

## Where should the reviewer start?

Checking the new code added into `lib/paper_trail.ex` and `lib/paper_trail/version_queries.ex`.

### CODE REVIEW

**Comment by izelnakri** [comment]
Thank you very much for such a high-quality pull request! I do not intend to support mysql however I can see that prefix feature could still be useful for postgres. Additionaly thanks for the documentation and code improvements.

If you could write a basic test case with two queries with prefixes and assert that those two produce different results influenced by prefixes then I can merge this pull request. We should run at least one test for this feature, so that it wont break somehow in future releases.

Thanks again for your contribution, feel free to add yourself to credits part as well ;)

**Comment by dreamingechoes** [comment]
Hi @izelnakri!

I'm glad to see that you found this PR interesting. I was waiting for your feedback to know if you agree with the changes I made and if I would have to make something else before add some tests for this, so I'll add them tomorrow so you can review the PR again and decide if you want to merge this.

Thank you so much for your awesome work with this package.

Greetings! 😄

**Comment by dreamingechoes** [comment]
Hi @izelnakri!

I just added some tests (in the `example` application tests as well as in the library ones) based on the original tests, so you can review it easily (or at least I hope so). 

Notice that to see the `example` on green, it's required to change the reference of the `paper_trail` package in the `example/mix.exs` file to the proper one, something like:

```
{:paper_trail, git: "https://github.com/dreamingechoes/paper_trail", branch: "add-multi-tenancy-capabilities-with-ecto-meta-prefix"}
```

Tell me if you see something weird in order to change it before merging the PR 😄

File: example/test/support/changeset_helper.exs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+defmodule ChangesetHelper do
```
Comment (izelnakri) [inline example/test/support/changeset_helper.exs]: I'm sorry to ask for this, but I think this abstraction makes are tests hard to read with very little benefit, could you remove this abstraction and change the tests?

File: example/test/support/query_helper.exs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+defmodule QueryHelper do
```
Comment (izelnakri) [inline example/test/support/query_helper.exs]: I'm sorry to ask for this, but I think this abstraction makes are tests hard to read with very little benefit, could you remove this abstraction and change the tests?

**Comment by dreamingechoes** [comment]
Hi @izelnakri!

You're right. I have created those modules in order to place the aux functions used in the `example` app tests and avoid the repetition of code, but it's true that this could make the tests harder to understand, so I just removed the extra modules and I have placed the aux functions right at the end of the test files so it could be more understandable.

Please if you have any other suggestion tell me, so I can make the changes and push them asap.

Thank you! 😄

File: example/test/company_test.exs [hunk 3]
```diff
@@ -164,4 +142,13 @@ defmodule CompanyTest do
       meta: nil
     }
   end
+
+  # Company related functions
+  def company_count(), do: (from company in Company, select: count(company.id))
+  def first_company(), do: (first(Company, :id) |> preload(:people))
+  def new_company(attrs), do: Company.changeset(%Company{}, attrs)
```
Comment (izelnakri) [inline example/test/company_test.exs]: new_company and updated_company functions could be removed for more explicitness. I think Company.changeset(company, attrs) is better than update_company(company, attrs)

File: example/test/person_test.exs [hunk 4]
```diff
@@ -9,46 +9,33 @@ defmodule PersonTest do
...
+    %{name: "Another Company Corp.", is_active: true, address: "Sesame street 100/3, 101010"}
+    |> new_company()
+    |> Repo.insert()
 
     :ok
   end
 
-  test "creating a person with meta tag creates a person version with correct attributes" do
-    company = first(Company, :id) |> Repo.one
+  test "[multi tenant] creating a person with meta tag creates a person version with correct attributes" do
```
Comment (izelnakri) [inline example/test/person_test.exs]: this doesnt test mulit-tenancy?

Comment (izelnakri) [inline example/test/person_test.exs]: We are infact never testing multi tenancy in this file as far as I see, (Im not suggesting that we should)

Comment (dreamingechoes) [inline example/test/person_test.exs]: Omg, this is embarrassing 😳 It's a typo due some "copy/paste" on the test descriptions, I'll remove this.

File: test/paper_trail/paper_trail_version_test.exs [hunk 5]
```diff
@@ -63,7 +71,30 @@ defmodule PaperTrailTest.Version do
     }
   end
 
-  def add_three_versions do
+  # Multi tenant tests
+  test "[multi tenant] count works" do
+    versions = add_three_versions(MultiTenant.tenant())
+    Version.count(prefix: MultiTenant.tenant()) == length(versions)
```
Comment (izelnakri) [inline test/paper_trail/paper_trail_version_test.exs]: we should also add that Version.count() != length(versions)

File: test/paper_trail/paper_trail_version_test.exs [hunk 6]
```diff
@@ -63,7 +71,30 @@ defmodule PaperTrailTest.Version do
...
+  end
+
+  test "[multi tenant] first works" do
+    add_three_versions(MultiTenant.tenant())
+    Version.first(prefix: MultiTenant.tenant()) |> serialize == @valid_attrs
```
Comment (izelnakri) [inline test/paper_trail/paper_trail_version_test.exs]: we should also add: Version.first() |> serialize != @valid_attrs

File: test/paper_trail/paper_trail_version_test.exs [hunk 7]
```diff
@@ -63,7 +71,30 @@ defmodule PaperTrailTest.Version do
...
+  test "[multi tenant] last works" do
+    add_three_versions(MultiTenant.tenant())
+    Version.last(prefix: MultiTenant.tenant()) |> serialize != %{
+      event: "insert",
+      item_type: "Person",
+      item_id: 3,
+      item_changes: %{first_name: "Yukihiro", last_name: "Matsumoto"},
+      origin: "test",
+      inserted_at: DateTime.from_naive!(~N[1965-04-14 01:00:00.000], "Etc/UTC")
+    }
```
Comment (izelnakri) [inline test/paper_trail/paper_trail_version_test.exs]: we should also add: Version.last() == nil

File: test/paper_trail/version_queries_test.exs [hunk 8]
```diff
@@ -61,26 +61,65 @@ defmodule PaperTrailTest.VersionQueries do
...
+      |> MultiTenant.add_prefix_to_query()
+      |> @repo.one
+    target_version_multi = last(Version, :id)
+      |> MultiTenant.add_prefix_to_query()
+      |> @repo.one
 
     assert PaperTrail.get_version(last_person) == target_version
     assert PaperTrail.get_version(Person, last_person.id) == target_version
+    assert PaperTrail.get_version(last_person_multi, prefix: tenant) == target_version_multi
+    assert PaperTrail.get_version(Person, last_person_multi.id, prefix: tenant) == target_version_multi
```
Comment (izelnakri) [inline test/paper_trail/version_queries_test.exs]: also target_version != target_version_multi

File: test/paper_trail/version_queries_test.exs [hunk 9]
```diff
@@ -61,26 +61,65 @@ defmodule PaperTrailTest.VersionQueries do
...
+      |> @repo.one
+    target_versions_multi = from(version in Version,
+      where: version.item_type == "SimplePerson" and version.item_id == ^last_person_multi.id)
+      |> MultiTenant.add_prefix_to_query()
+      |> @repo.all
 
     assert PaperTrail.get_versions(last_person) == target_versions
     assert PaperTrail.get_versions(Person, last_person.id) == target_versions
+    assert PaperTrail.get_versions(last_person_multi, prefix: tenant) == target_versions_multi
+    assert PaperTrail.get_versions(Person, last_person_multi.id, prefix: tenant) == target_versions_multi
```
Comment (izelnakri) [inline test/paper_trail/version_queries_test.exs]: also target_version != target_version_multi

File: test/paper_trail/bang_functions_strict_mode_test.exs [hunk 10]
```diff
@@ -388,24 +386,470 @@ defmodule PaperTrailTest.StrictModeBangFunctions do
...
+        birthdate: elem(Ecto.Date.cast(~D[1992-04-01]), 1),
+        company_id: inserted_company.id,
+        first_version_id: inserted_person_version.id,
+        current_version_id: updated_person_version.id
+      }),
+      originator_id: nil,
+      origin: "admin",
+      meta: %{"linkname" => "izelnakri"}
+    }
+    assert deleted_person |> serialize == person_before_deletion |> serialize
```
Comment (izelnakri) [inline test/paper_trail/bang_functions_strict_mode_test.exs]: also assert that Version.count() == 0

File: test/paper_trail/bang_functions_strict_mode_test.exs [hunk 11]
```diff
@@ -388,24 +386,470 @@ defmodule PaperTrailTest.StrictModeBangFunctions do
...
+        visit_count: 10,
+        birthdate: elem(Ecto.Date.cast(~D[1992-04-01]), 1),
+        current_version_id: updated_person_version.id,
+        company_id: inserted_initial_company.id
+      }),
+      originator_id: nil,
+      origin: "scraper",
+      meta: %{"linkname" => "izelnakri"}
+    }
+    assert person == first_person(:multitenant) |> serialize
```
Comment (izelnakri) [inline test/paper_trail/bang_functions_strict_mode_test.exs]: also assert that Version.count() == 0

File: example/test/multi_tenant_company_test.exs [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+      |> preload(:people)
+      |> MultiTenantHelper.add_prefix_to_query()
+      |> Repo.one()
+
+    company = result[:model] |> Map.drop([:__meta__, :__struct__, :inserted_at, :updated_at, :id])
+    version = result[:version] |> Map.drop([:__meta__, :__struct__, :inserted_at])
+
+    assert company_count == [1]
+    assert version_count == [1]
+
```
Comment (izelnakri) [inline example/test/multi_tenant_company_test.exs]: Version.count == [0]

File: example/test/multi_tenant_company_test.exs [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+      from(version in PaperTrail.Version, select: count(version.id))
+      |> MultiTenantHelper.add_prefix_to_query()
+      |> Repo.all()
+
+    company = result[:model] |> Map.drop([:__meta__, :__struct__, :inserted_at, :updated_at, :id])
+    version = result[:version] |> Map.drop([:__meta__, :__struct__, :inserted_at])
+
+    assert company_count == [1]
+    assert version_count == [2]
+
```
Comment (izelnakri) [inline example/test/multi_tenant_company_test.exs]: assert Version.count == [0]

File: example/test/multi_tenant_company_test.exs [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+      from(version in PaperTrail.Version, select: count(version.id))
+      |> MultiTenantHelper.add_prefix_to_query()
+      |> Repo.all()
+
+    company_ref = result[:model] |> Map.drop([:__meta__, :__struct__, :inserted_at, :updated_at, :id])
+    version = result[:version] |> Map.drop([:__meta__, :__struct__, :inserted_at])
+
+    assert company_count == [0]
+    assert version_count == [3]
+
```
Comment (izelnakri) [inline example/test/multi_tenant_company_test.exs]: assert Version.count == [0]

**Comment by izelnakri** [comment]
Fantastic work, thank you very much!

**Comment by dreamingechoes** [comment]
Thanks to you @izelnakri! I'm glad to contribute to this package. Thank you for your work and your awesome feedback during the pull request 😄

**Pull request merged**